### PR TITLE
Add correct batched handling for apply_chat_template

### PR DESCRIPTION
**Pull request by Rocketknight1**
`apply_chat_template` has a few issues since it was written. Firstly, by default it returns the naked `input_ids` rather than a dict, and secondly it didn't support rendering a batch of chats simultaneously. This PR makes a few changes:

- Batched chats are now supported, and we sniff the input to figure out what the user is passing
- `return_dict` now defaults to `None`. For now, we interpret this as `False` to maintain backward compatibility, but this PR adds a warning that the default behaviour will be changing to `True` to match other tokenizer methods.

cc @siddk @lewtun who have both requested this or something like it!

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Should be ready for review now! cc @ArthurZucker

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -1719,7 +1719,7 @@ def apply_chat_template(
                 - `'pt'`: Return PyTorch `torch.Tensor` objects.
                 - `'np'`: Return NumPy `np.ndarray` objects.
                 - `'jax'`: Return JAX `jnp.ndarray` objects.
-            return_dict (`bool`, *optional*, defaults to `False`):
+            return_dict (`bool`, *optional*, defaults to `None`):
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1722]:
```suggestion
            return_dict (`bool`, *optional*):
```

File: src/transformers/tokenization_utils_base.py [hunk 2]
```diff
@@ -1728,9 +1728,16 @@ def apply_chat_template(
...
 
-        if hasattr(conversation, "messages"):
-            # Indicates it's a Conversation object
-            conversation = conversation.messages
+        if return_dict is None and tokenize:
+            logger.warning_once(
+                "In version 4.40, `return_dict` will be set to `True` by default. "
+                "Please explicitly set `return_dict` to `False` to maintain the current behaviour, "
+                "or set it to `True` to get the new behaviour immediately."
+            )
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1736]: would be nice to explain why this should be set to `True` for example? I have no idea

File: src/transformers/tokenization_utils_base.py [hunk 3]
```diff
@@ -1728,9 +1728,16 @@ def apply_chat_template(
...
+            return_dict = False
+
+        if return_dict and not tokenize:
+            raise ValueError("return_dict=True is incompatible with tokenize=False")
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1740]:
```suggestion
            raise ValueError("`return_dict=True` is incompatible with `tokenize=False`")
```
also explaining why should help users

File: src/transformers/tokenization_utils_base.py [hunk 4]
```diff
@@ -1677,17 +1677,17 @@ def get_vocab(self) -> Dict[str, int]:
...
         truncation: bool = False,
         max_length: Optional[int] = None,
         return_tensors: Optional[Union[str, TensorType]] = None,
-        return_dict: bool = False,
+        return_dict: bool = None,
         **tokenizer_kwargs,
-    ) -> Union[str, List[int]]:
+    ) -> Union[str, List[int], List[str], List[List[int]], BatchEncoding]:
         """
         Converts a Conversation object or a list of dictionaries with `"role"` and `"content"` keys to a list of token
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1692]: docstring for `conversation` needs to be updated as well

File: src/transformers/tokenization_utils_base.py [hunk 5]
```diff
@@ -1719,7 +1719,7 @@ def apply_chat_template(
                 - `'pt'`: Return PyTorch `torch.Tensor` objects.
                 - `'np'`: Return NumPy `np.ndarray` objects.
                 - `'jax'`: Return JAX `jnp.ndarray` objects.
-            return_dict (`bool`, *optional*, defaults to `False`):
+            return_dict (`bool`, *optional*, defaults to `None`):
                 Whether to return a dictionary with named outputs. Has no effect if tokenize is `False`.
             **tokenizer_kwargs: Additional kwargs to pass to the tokenizer.
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1725]: return needs to be updated

File: src/transformers/tokenization_utils_base.py [hunk 6]
```diff
@@ -1742,33 +1749,44 @@ def apply_chat_template(
...
         compiled_template = self._compile_jinja_template(chat_template)
 
-        rendered = compiled_template.render(
-            messages=conversation, add_generation_prompt=add_generation_prompt, **self.special_tokens_map
-        )
+        if isinstance(conversation, (list, tuple)) and (
+            isinstance(conversation[0], (list, tuple)) or hasattr(conversation[0], "messages")
+        ):
+            conversations = conversation
+            batched = True
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1756]:
```suggestion
            it_batched = True
```

File: src/transformers/tokenization_utils_base.py [hunk 7]
```diff
@@ -1742,33 +1749,44 @@ def apply_chat_template(
...
+        else:
+            batched = False
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1758]:
```suggestion
            it_batched = False
```

File: src/transformers/tokenization_utils_base.py [hunk 8]
```diff
@@ -1742,33 +1749,44 @@ def apply_chat_template(
...
+        for chat in conversations:
+            if hasattr(chat, "messages"):
+                # Indicates it's a Conversation object
+                chat = chat.messages
+
+            rendered.append(
+                compiled_template.render(
+                    messages=chat, add_generation_prompt=add_generation_prompt, **self.special_tokens_map
+                )
+            )
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1771]:
```suggestion
            rendered_chat = compiled_template.render(messages=chat, add_generation_prompt=add_generation_prompt, **self.special_tokens_map)
            rendered.append(rendered_chat)
```

File: src/transformers/tokenization_utils_base.py [hunk 9]
```diff
@@ -1742,33 +1749,44 @@ def apply_chat_template(
...
+
+        if not batched:
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1773]:
```suggestion
        if not is_batched:
```

File: src/transformers/tokenization_utils_base.py [hunk 10]
```diff
@@ -1742,33 +1749,44 @@ def apply_chat_template(
...
+            rendered = rendered[0]
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1774]: should we not always return a batched output? (breaking but we can warn)

File: src/transformers/tokenization_utils_base.py [hunk 11]
```diff
@@ -1742,33 +1749,44 @@ def apply_chat_template(
...
         if tokenize:
+            out = self(
+                rendered,
+                padding=padding,
+                truncation=truncation,
+                max_length=max_length,
+                add_special_tokens=False,
+                return_tensors=return_tensors,
+                **tokenizer_kwargs,
+            )
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1785]: nice

**Review by ArthurZucker — APPROVED** [review]
LGTM !

File: src/transformers/tokenization_utils_base.py (hunk 4 above)
Comment (Rocketknight1) [reply src/transformers/tokenization_utils_base.py]: Updated!

File: src/transformers/tokenization_utils_base.py (hunk 5 above)
Comment (Rocketknight1) [reply src/transformers/tokenization_utils_base.py:1725]: Done!

File: src/transformers/tokenization_utils_base.py (hunk 10 above)
Comment (Rocketknight1) [reply src/transformers/tokenization_utils_base.py:1771]: I'm not sure - other tokenizer methods don't auto-batch a single input, right? (And sorry for taking so long to reply here!)

File: src/transformers/tokenization_utils_base.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/tokenization_utils_base.py]: I changed my mind about this and removed the warning to make this a simpler PR!

**Comment by Rocketknight1** [comment]
This should be ready for re-review now, cc @amyeroberts @ArthurZucker! I simplified the PR by removing the deprecation warning - I'm not sure if we want to move to `return_dict=True` that quickly anyway. As a result, this shouldn't result in any behaviour changes now, it only adds new functionality.

File: src/transformers/tokenization_utils_base.py [hunk 12]
```diff
@@ -1730,18 +1730,24 @@ def apply_chat_template(
...
+            return_dict (`bool`, *optional*):
                 Whether to return a dictionary with named outputs. Has no effect if tokenize is `False`.
             **tokenizer_kwargs: Additional kwargs to pass to the tokenizer.
 
         Returns:
-            `List[int]`: A list of token ids representing the tokenized chat so far, including control tokens. This
-            output is ready to pass to the model, either directly or via methods like `generate()`.
+            `Union[List[int], Dict]`: A list of token ids representing the tokenized chat so far, including control tokens. This
+            output is ready to pass to the model, either directly or via methods like `generate()`. If `return_dict` is
+            set, will return a dict of tokenizer output instead.
```
Comment (amyeroberts) [inline src/transformers/tokenization_utils_base.py:1740]: nit - I think? Grammar is not my strong suit 

```suggestion
            set, will return a dict of tokenizer outputs instead.
```

File: src/transformers/tokenization_utils_base.py [hunk 13]
```diff
@@ -1753,33 +1759,42 @@ def apply_chat_template(
...
-            messages=conversation, add_generation_prompt=add_generation_prompt, **self.special_tokens_map
-        )
+        if isinstance(conversation, (list, tuple)) and (
+            isinstance(conversation[0], (list, tuple)) or hasattr(conversation[0], "messages")
+        ):
+            conversations = conversation
+            is_batched = True
+        else:
+            is_batched = False
+            conversations = [conversation]
```
Comment (amyeroberts) [inline src/transformers/tokenization_utils_base.py:1769]: ultra ultra nit - it's nice to have the logic steps in the same order for each part of the if/else branch

```suggestion
            conversations = [conversation]
            is_batched = False
```

File: tests/test_tokenization_common.py [hunk 14]
```diff
@@ -1102,26 +1102,75 @@ def test_chat_template(self):
...
 
                 tokenizer.chat_template = dummy_template
                 self.assertEqual(tokenizer.chat_template, dummy_template)  # Test property setter
-                output = tokenizer.apply_chat_template(dummy_conversation, tokenize=False)
+                output = tokenizer.apply_chat_template(dummy_conversation, tokenize=False, return_dict=False)
                 self.assertEqual(output, expected_output)  # Test chat_template attribute is used if no arg is passed
-                tokenizer.apply_chat_template(dummy_conversation, tokenize=True)  # Check that no error raised
+                tokenizer.apply_chat_template(
+                    dummy_conversation, tokenize=True, return_dict=False
+                )  # Check that no error raised
```
Comment (amyeroberts) [inline tests/test_tokenization_common.py:1124]: nit - comment above means we use fewer lines

```suggestion
                # Check that no error raised
                tokenizer.apply_chat_template(dummy_conversation, tokenize=True, return_dict=False)
```

File: tests/test_tokenization_common.py [hunk 15]
```diff
@@ -1102,26 +1102,75 @@ def test_chat_template(self):
...
                     tokenizer = tokenizer.from_pretrained(tmp_dir_name)
 
                 self.assertEqual(tokenizer.chat_template, dummy_template)  # Test template has persisted
-                output = tokenizer.apply_chat_template(dummy_conversation, tokenize=False)
+                output = tokenizer.apply_chat_template(dummy_conversation, tokenize=False, return_dict=False)
                 self.assertEqual(output, expected_output)  # Test output is the same after reloading
-                tokenizer.apply_chat_template(dummy_conversation, tokenize=True)  # Check that no error raised
+                tokenizer.apply_chat_template(
+                    dummy_conversation, tokenize=True, return_dict=False
+                )  # Check that no error raised
```
Comment (amyeroberts) [inline tests/test_tokenization_common.py:1135]:
```suggestion
                # Check that no error raised
                tokenizer.apply_chat_template(dummy_conversation, tokenize=True, return_dict=False)
```

File: src/transformers/tokenization_utils_base.py [hunk 16]
```diff
@@ -1730,18 +1730,24 @@ def apply_chat_template(
                 - `'pt'`: Return PyTorch `torch.Tensor` objects.
                 - `'np'`: Return NumPy `np.ndarray` objects.
                 - `'jax'`: Return JAX `jnp.ndarray` objects.
-            return_dict (`bool`, *optional*, defaults to `False`):
+            return_dict (`bool`, *optional*):
```
Comment (amyeroberts) [inline src/transformers/tokenization_utils_base.py:1733]: Why change the default to `None` here? AFAICT, this doesn't change things. It gets set to `False` if `tokenize` is True, but it's only used in truth checks on L1763 and L1773 (which shouldn't really do this if the value can be none anyway) and `False` or `None` will have the same result

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this! 

Two things before merge: 
* Question about the default value of `return_dict` 
* Let's wait for @ArthurZucker to get back to confirm desired batching behaviour

File: src/transformers/tokenization_utils_base.py (hunk 12 above)
Comment (Rocketknight1) [reply src/transformers/tokenization_utils_base.py]: You're right! I think `output` would only be right if we were treating it as an uncountable mass noun, which I'm pretty sure we aren't.

File: src/transformers/tokenization_utils_base.py (hunk 16 above)
Comment (Rocketknight1) [reply src/transformers/tokenization_utils_base.py:1733]: Fixed! You're right - this is leftovers from when I was planning to slowly make `return_dict=True` the default.

**Comment by Rocketknight1** [comment]
Merging this now that the branch cut has passed!

**Pull request merged**